src/Chapter3.hs

```diff
@@ -1005,7 +1130,98 @@ properties using typeclasses, but they are different data types in the end.
 Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
-
+newtype Potion = Potion Int
+newtype Spell = Spell Int
+
+class Fighter a where
+  attack :: a -> a -> a
+  isAttacked :: Int -> a -> a
+  runAway :: a -> a
+  health :: a -> Int
+  -- act :: a -> a
+  -- endTurn :: a -> a
+
+-- data KnightAction = KAttack | KRunAway | KDrinkPotion | KCastSpell
+data KnightT9 = KnightT9
+  { knightAttack :: Int
+  , knightHealth :: Int
+  , knightDefense :: Int
+  -- , knightActions :: [KnightAction]
```

Comment:
You are on the right track of having separate types `KnightAction` and `MonsterAction`. Those should be separate types.

This task has multiple solutions and multiple ways of solving the problem.

Just a few ideas:

1. Introduce a function: `getCurrentAction :: a -> Either KnightAction MonsterAction`
2. Have more modular typeclasses instead of a uniform `Fighter`.
3. Introduce a uniform method: `performCurrentAction :: a -> NewStatus a` where this function will perform current action as a part of the implementation and then return new status of a fighter: are they run away, are they drink potion, or else? So, with this approach you don't need to worry about handling the type of action as a part of a typeclass.

I would vote for option 3

Reply:
Thanks for the feedback @chshersh! I'll try to implement this and have a new, proposed version in the next few hours.

Reply:
@DavSanchez Sure! No worries :slightly_smiling_face: I see that you've already completed other chapters. Let us know if you want any feedback on this chapter. Otherwise, feel free to merge whenever you like :slightly_smiling_face:

Reply:
Hi @chshersh!

I have uploaded a new version that is still WIP, but I'd like to get some feedback about it.

I'm still struggling with the `attack` action, as it requires knowing the other fighter in order to modify its stats. This conflicts with the `performAction` function, which returns the same `Fighter` that called it. Should the other methods (`runAway`, `drinkPotion`, `castDefensiveSpell`) take the rival as well for the function to be uniform? Perhaps there should be a type like `Battle {fighter1 :: Fighter, fighter2 :: Fighter}`? I don't think that's even possible in Haskell. Perhaps a sum type with the four combinations...

I'm working with the idea of calling `performAction` on the first argument of `fightT9` (the very last function of the file) and then calling `fightT9` recursively by swapping the arguments (imitating the consecutive turns of a fight).